Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/TrainControlSystem.cs

```diff
@@ -670,9 +670,10 @@ SignalFeatures NextGenericSignalFeatures(string signalFunctionTypeName, int item
                     }
 
                     var functionHead = default(SignalHead);
-                    foreach (var head in trainSignal.SignalObject.SignalHeads)
-                        if (head.Function == function)
-                            functionHead = head;
+                    if (trainSignal?.SignalObject?.SignalHeads != null)
```

Comment:
Adding a null check is just hiding the real problem. Those structures should never be null, so there is no point in adding null pointer checks. If they are null, that's due to a corrupted signalling in the route. Maybe we should be able to recover from that instead of crashing, but that should be done by ensuring the consistency of the signalling at initialization.

Reply:
Yeah, you are right, but... They admitted that the signal scripts may be wrong, but they are unable to debug it, and the original creator passed away. And I think it is never a good approach to run the game onto a nullreferenceexception. Yes, an initial check could be added, and the problem logged, but still, if the script is wrong, we should not run on a nullreferenceexception. That's why all the sript functions are written in a way to return at least an empty string, not a null, to avoid this.

Reply:
I agree that it's wrong to throw the exception. The thing is that it makes no sense to me that those structures are null. I think this started with the refactoring in #537, previous code was more tolerant to badly configured SIGCFGs. We should ensure that they are not null instead of adding null pointer checks everywhere, which makes the code much less readable. Maybe it is enough to check for null at initialization, and then default-initialize SignalHeads, SignalType and so on.